File: app/src/main/java/com/tasomaniac/openwith/AppComponent.java
package com.tasomaniac.openwith;

import com.tasomaniac.openwith.data.Analytics;
import com.tasomaniac.openwith.resolver.ResolverBindingModule;
import com.tasomaniac.openwith.resolver.ResolverComponent;

import javax.inject.Singleton;

import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(modules = {
        AndroidSupportInjectionModule.class,
        AppModule.class,
        AnalyticsModule.class,
        BindingModule.class,
        ResolverBindingModule.class
})
public interface AppComponent extends AndroidInjector<App> {

    ResolverComponent.Builder resolverComponentBuilder();
    Analytics analytics();

    @Component.Builder
    abstract class Builder extends AndroidInjector.Builder<App> {

        @Override
        public abstract AppComponent build();
    }
}

File: app/src/main/java/com/tasomaniac/openwith/AppComponent.kt
package com.tasomaniac.openwith

import com.tasomaniac.openwith.data.Analytics
import com.tasomaniac.openwith.data.DataModule
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        AndroidSupportInjectionModule::class,
        AppModule::class,
        AnalyticsModule::class,
        DataModule::class,
        BindingModule::class
    ]
)
interface AppComponent : AndroidInjector<App> {

    fun analytics(): Analytics

    @Component.Factory
    interface Factory : AndroidInjector.Factory<App> {
        override fun create(@BindsInstance instance: App): AppComponent
    }
}
